File: extension/src/lib/classify.ts
// Status classification, mirroring the CLI's cve/classify.go: database first, live NVD
// as fallback, Retire.js merged in, release cycles deciding between current/outdated/EOL.

import type { Cache } from "./cache.ts";
import { cpeWithVersion, productFromCpeName } from "./cpe.ts";
import {
  databaseVulnerabilities,
  libraryVulnerabilities,
  lifecycleFor,
  mergeVulnerabilities,
  productFor,
  type Database,
} from "./db.ts";
import { lookupKev, type KevCatalog } from "./kev.ts";
import type { NvdClient } from "./nvd.ts";
import type { Assessment, LifecycleInfo, Status, Technology, Vulnerability } from "./types.ts";
import { parseVersion, type Version } from "./version.ts";

const RANK: Record<Status, number> = {
  current: 0,
  outdated: 1,
  unknown: 2,
  unsupported: 3,
  vulnerable: 4,
  critical: 5,
};

/** Orders statuses by urgency, from current (0) to critical (5). */
export function rank(status: Status): number {
  return RANK[status];
}

/**
 * Returns the parsed version or the reason no verdict is possible. Single-segment
 * versions such as "8" match every 8.x advisory and would produce false alarms.
 */
export function comparableVersion(
  technology: Technology,
): { version: Version } | { reason: string } {
  if (!technology.version) {
    return { reason: "version not disclosed" };
  }
  if (!technology.version.includes(".")) {
    return { reason: `version "${technology.version}" is too coarse to match advisories` };
  }
  const version = parseVersion(technology.version);
  if (!version) {
    return { reason: `version "${technology.version}" is not comparable` };
  }
  return { version };
}

function moreUrgent(a: Vulnerability, b: Vulnerability): number {
  if ((a.kev !== undefined) !== (b.kev !== undefined)) {
    return a.kev ? -1 : 1;
  }
  const aExploit = (a.exploits?.length ?? 0) > 0;
  const bExploit = (b.exploits?.length ?? 0) > 0;
  if (aExploit !== bExploit) {
    return aExploit ? -1 : 1;
  }
  const epssDiff = (b.epss ?? 0) - (a.epss ?? 0);
  if (epssDiff !== 0) {
    return epssDiff;
  }
  const scoreDiff = (b.score ?? 0) - (a.score ?? 0);
  if (scoreDiff !== 0) {
    return scoreDiff;
  }
  return b.id.localeCompare(a.id);
}

/**
 * Derives a status from applicable vulnerabilities, the KEV catalog and release data.
 * Vulnerabilities are ordered exploited in the wild → public exploit → EPSS → CVSS.
 */
export function classify(
  technology: Technology,
  cpeName: string,
  vulnerabilities: Vulnerability[],
  kev: KevCatalog | null,
  lifecycle: LifecycleInfo | null = null,
): Assessment {
  const assessment: Assessment = { technology, status: "current" };
  if (cpeName) assessment.cpe_name = cpeName;
  if (lifecycle) assessment.lifecycle = lifecycle;
  if (vulnerabilities.length === 0) {
    if (lifecycle?.eol) {
      assessment.status = "unsupported";
    } else if (lifecycle?.behind) {
      assessment.status = "outdated";
    }
    return assessment;
  }
  assessment.status = "vulnerable";
  assessment.vulnerabilities = vulnerabilities.map((vulnerability) => {
    const entry = vulnerability.kev ?? lookupKev(kev, vulnerability.id);
    if (!entry) {
      return { ...vulnerability };
    }
    assessment.status = "critical";
    return { ...vulnerability, kev: entry };
  });
  assessment.vulnerabilities.sort(moreUrgent);
  return assessment;
}

export class Assessor {
  constructor(
    private readonly nvd: NvdClient | null,
    private readonly kev: KevCatalog | null,
    private readonly cache: Cache | null,
    private readonly database: Database | null = null,
  ) {}

  /** Looks up CVEs and release status for a technology; lookup failures yield "unknown". */
  async assess(technology: Technology): Promise<Assessment> {
    const parsed = comparableVersion(technology);
    if ("reason" in parsed) {
      return { technology, status: "unknown", reason: parsed.reason };
    }
    const record = productFor(this.database, technology.cpe);
    const libraryKey = this.database?.technology_libraries[technology.name];
    const hasLibrary =
      libraryKey !== undefined && this.database?.libraries[libraryKey] !== undefined;
    if (!technology.cpe && !hasLibrary) {
      return { technology, status: "unknown", reason: "no CPE mapping for this technology" };
    }
    let vulnerabilities: Vulnerability[] = [];
    let lifecycle: LifecycleInfo | null = null;
    let cpeName = "";
    if (technology.cpe) {
      try {
        cpeName = cpeWithVersion(technology.cpe, technology.version ?? "");
        const target = productFromCpeName(cpeName);
        if (record && this.database) {
          vulnerabilities = databaseVulnerabilities(this.database, record, target);
          lifecycle = lifecycleFor(record, parsed.version, technology.version ?? "");
        } else if (this.nvd) {
          vulnerabilities = await this.liveVulnerabilities(cpeName);
        } else if (!hasLibrary) {
          return {
            technology,
            status: "unknown",
            reason: "not in the vulnerability database",
            cpe_name: cpeName,
          };
        }
      } catch (error) {
        const reason = error instanceof Error ? error.message : String(error);
        const failed: Assessment = { technology, status: "unknown", reason };
        if (cpeName) failed.cpe_name = cpeName;
        return failed;
      }
    }
    if (hasLibrary) {
      vulnerabilities = mergeVulnerabilities(
        vulnerabilities,
        libraryVulnerabilities(this.database, technology.name, parsed.version),
      );
    }
    return classify(technology, cpeName, vulnerabilities, this.kev, lifecycle);
  }

  private async liveVulnerabilities(cpeName: string): Promise<Vulnerability[]> {
    const cached = await this.cache?.get<Vulnerability[]>(cpeName);
    if (cached) {
      return cached;
    }
    if (!this.nvd) {
      return [];
    }
    const fresh = await this.nvd.vulnerabilitiesFor(cpeName);
    await this.cache?.set(cpeName, fresh);
    return fresh;
  }
}

File: cve/classify.go
package cve

import (
	"context"
	"fmt"
	"sort"
	"strings"

	"github.com/hashicorp/go-version"
	"github.com/xZoroo/wappacvelyze/db"
	"github.com/xZoroo/wappacvelyze/detect"
)

// Status is the security verdict for a detected technology version.
type Status string

const (
	// StatusCurrent means no CVE applies and, when release data exists, the version is the
	// newest in a maintained cycle.
	StatusCurrent Status = "current"
	// StatusOutdated means no CVE applies but a newer release exists in the same cycle.
	StatusOutdated Status = "outdated"
	// StatusUnknown means no lookup was possible; Assessment.Reason says why.
	StatusUnknown Status = "unknown"
	// StatusUnsupported means the release cycle is end-of-life, so fixes will not come.
	StatusUnsupported Status = "unsupported"
	// StatusVulnerable means at least one CVE applies to the detected version.
	StatusVulnerable Status = "vulnerable"
	// StatusCritical means an applicable CVE is on CISA's Known Exploited Vulnerabilities list.
	StatusCritical Status = "critical"
)

var statusRank = map[Status]int{
	StatusCurrent: 0, StatusOutdated: 1, StatusUnknown: 2,
	StatusUnsupported: 3, StatusVulnerable: 4, StatusCritical: 5,
}

// Rank orders statuses by urgency, from current (0) to critical (5).
func (s Status) Rank() int {
	return statusRank[s]
}

// ParseStatus accepts a status name as typed on the command line.
func ParseStatus(s string) (Status, error) {
	status := Status(strings.ToLower(s))
	if _, ok := statusRank[status]; ok {
		return status, nil
	}
	return "", fmt.Errorf("unknown status %q (want current, outdated, unknown, unsupported, vulnerable or critical)", s)
}

// Assessment is the security status of one detected technology.
type Assessment struct {
	Technology      detect.Technology `json:"technology"`
	Status          Status            `json:"status"`
	Reason          string            `json:"reason,omitempty"`
	CPEName         string            `json:"cpe_name,omitempty"`
	Lifecycle       *LifecycleInfo    `json:"lifecycle,omitempty"`
	Vulnerabilities []Vulnerability   `json:"vulnerabilities,omitempty"`
}

// Assessor resolves detected technologies to security statuses. It prefers the prebuilt
// database and falls back to live NVD lookups for products the database lacks. Any field
// may be nil.
type Assessor struct {
	DB    *db.Database
	NVD   *NVDClient
	KEV   *KEVCatalog
	Cache *Cache
}

// Assess looks up CVEs and release status for t. When a live lookup fails the assessment
// is returned as StatusUnknown alongside the error so callers can report both.
func (a *Assessor) Assess(ctx context.Context, t detect.Technology) (Assessment, error) {
	detected, reason := comparableVersion(t)
	if reason != "" {
		return Assessment{Technology: t, Status: StatusUnknown, Reason: reason}, nil
	}
	var vulns []Vulnerability
	var lifecycle *LifecycleInfo
	cpeName := ""
	entry := databaseProduct(a.DB, t.CPE)
	library := a.DB != nil && a.DB.Libraries[a.DB.TechnologyLibraries[t.Name]] != nil
	switch {
	case t.CPE == "" && !library:
		return Assessment{Technology: t, Status: StatusUnknown, Reason: "no CPE mapping for this technology"}, nil
	case t.CPE != "":
		name, err := CPEWithVersion(t.CPE, t.Version)
		if err != nil {
			return Assessment{Technology: t, Status: StatusUnknown, Reason: err.Error()}, nil
		}
		cpeName = name
		target, err := productFromCPEName(cpeName)
		if err != nil {
			return Assessment{Technology: t, Status: StatusUnknown, Reason: err.Error()}, nil
		}
		if entry != nil {
			vulns = databaseVulnerabilities(a.DB, entry, target)
			lifecycle = lifecycleFor(entry, detected, t.Version)
		} else if a.NVD != nil {
			vulns, err = a.vulnerabilities(ctx, cpeName)
			if err != nil {
				failed := Assessment{Technology: t, Status: StatusUnknown, Reason: "lookup failed", CPEName: cpeName}
				return failed, err
			}
		} else if !library {
			return Assessment{Technology: t, Status: StatusUnknown, Reason: "not in the vulnerability database", CPEName: cpeName}, nil
		}
	}
	if library {
		vulns = mergeVulnerabilities(vulns, libraryVulnerabilities(a.DB, t.Name, detected))
	}
	return Classify(t, cpeName, vulns, a.KEV, lifecycle), nil
}

func (a *Assessor) vulnerabilities(ctx context.Context, cpeName string) ([]Vulnerability, error) {
	if a.Cache != nil {
		if vulns, ok := a.Cache.Get(cpeName); ok {
			return vulns, nil
		}
	}
	vulns, err := a.NVD.VulnerabilitiesFor(ctx, cpeName)
	if err != nil {
		return nil, err
	}
	if a.Cache != nil {
		if err := a.Cache.Put(cpeName, vulns); err != nil {
			return nil, err
		}
	}
	return vulns, nil
}

// comparableVersion returns the parsed version or the reason no verdict is possible.
// Single-segment versions such as "8" match every 8.x advisory and would produce false
// alarms, so they are treated as undisclosed.
func comparableVersion(t detect.Technology) (*version.Version, string) {
	switch {
	case t.Version == "":
		return nil, "version not disclosed"
	case !strings.Contains(t.Version, "."):
		return nil, fmt.Sprintf("version %q is too coarse to match advisories", t.Version)
	}
	detected, err := parseVersion(t.Version)
	if err != nil {
		return nil, err.Error()
	}
	return detected, ""
}

// Classify derives a status from applicable vulnerabilities, the KEV catalog and release
// data. Vulnerabilities are ordered most urgent first: exploited in the wild, then with a
// public exploit, then by EPSS probability, then by CVSS score.
func Classify(
	t detect.Technology, cpeName string, vulns []Vulnerability, kev *KEVCatalog, lifecycle *LifecycleInfo,
) Assessment {
	a := Assessment{Technology: t, Status: StatusCurrent, CPEName: cpeName, Lifecycle: lifecycle}
	if len(vulns) == 0 {
		switch {
		case lifecycle != nil && lifecycle.EOL:
			a.Status = StatusUnsupported
		case lifecycle != nil && lifecycle.Behind:
			a.Status = StatusOutdated
		}
		return a
	}
	a.Status = StatusVulnerable
	a.Vulnerabilities = make([]Vulnerability, len(vulns))
	for i, v := range vulns {
		if v.KEV == nil && kev != nil {
			if entry, ok := kev.Lookup(v.ID); ok {
				v.KEV = &entry
			}
		}
		if v.KEV != nil {
			a.Status = StatusCritical
		}
		a.Vulnerabilities[i] = v
	}
	sort.SliceStable(a.Vulnerabilities, func(i, j int) bool {
		return moreUrgent(a.Vulnerabilities[i], a.Vulnerabilities[j])
	})
	return a
}

func moreUrgent(a, b Vulnerability) bool {
	if (a.KEV != nil) != (b.KEV != nil) {
		return a.KEV != nil
	}
	if (len(a.Exploits) > 0) != (len(b.Exploits) > 0) {
		return len(a.Exploits) > 0
	}
	if a.EPSS != b.EPSS {
		return a.EPSS > b.EPSS
	}
	if a.Score != b.Score {
		return a.Score > b.Score
	}
	return a.ID > b.ID
}
